Add support for tags

playfulkittykat opened the pull request:
Here's my attempt at adding support for tags. It isn't entirely complete yet, but I wanted to get your opinion before moving forward.

I've gone with a builder pattern for `tags::Query`, but differently than `PoolSearch`. If you like it, I can update everything to the same style.

Outstanding Tasks:

 - [x] verify `Cursor` logic
 - [x] mockito tests
 - [x] better documentation for `tags::Query` (and everything else.)
 - [ ] more examples

playfulkittykat committed:
Add support for tags

playfulkittykat commented on src/tag.rs:
```diff
@@ -0,0 +1,415 @@
+            Ok(MaybeTags::Items(i)) => i,
+        };
+
+        let (min, max) = tags.iter().map(|x| x.id).minmax().into_option().unwrap();
+
+        let next_page = match query.page {
+            Some(Cursor::Before(_)) => Cursor::Before(min),
+            Some(Cursor::After(_)) => Cursor::After(max),
+            Some(Cursor::Page(p)) => Cursor::Page(p + 1),
+            None => match query.order {
+                None | Some(Order::IdDsc | Order::Date) => Cursor::Before(min),
+                Some(Order::IdAsc) => Cursor::After(max),
+                _ => Cursor::Page(2),
+            },
+        };
+        query.page = Some(next_page);
```
So this is the part I'm least sure about. I don't fully understand the whole `a1234` / `b1234` thing and how it relates to `search[order]`.

nasso commented:
thank you!

i like the `Cursor` and the `Query` API seems nice. we can work on making it more consistent with `PoolSearch` in a separate PR to make things easier

i am not familiar with the tag search API (it didn't exist back then) so i dont know much about the `a1234` and `b1234` bits

playfulkittykat commented:
> i like the `Cursor` and the `Query` API seems nice. we can work on making it more consistent with `PoolSearch` in a separate PR to make things easier

Thank you! That works for me

> i am not familiar with the tag search API (it didn't exist back then) so i dont know much about the `a1234` and `b1234` bits

It looks like it uses the same code path as the posts endpoint, so the `a...` and `b...` should be similar to:

https://github.com/nasso/rs621/blob/236bac4a4fc74e063ffbfc9718706e4c9c946ff7/src/post.rs#L388-L392

playfulkittykat committed:
Add mockito tests for /tags.json

playfulkittykat commented:
I'm happy with where this is now. Let me know what you think!

nasso commented on src/tag.rs:
````diff
@@ -0,0 +1,723 @@
+    /// }
+    /// # Ok(()) }
+    /// ```
+    #[cfg(not(any(target_arch = "wasm32", target_arch = "wasm64")))]
````
```suggestion
    #[cfg(not(target_family = "wasm"))]
```

same for the other ones

nasso commented on src/tag.rs:
```diff
@@ -0,0 +1,723 @@
+        };
+        query.page = Some(next_page);
+
+        // TODO: This allocates an unnecessary `Vec`.
+        let tag_results = tags.into_iter().map(Ok).collect();
```
can we avoid allocation?

nasso commented on src/tag.rs:
```diff
@@ -0,0 +1,723 @@
+                    "fish 69 hi_res 56 clothing 50 feral 49 male 49 female 48 ",
+                    "water 48 fur 47 text 44 hair 42 ambiguous_gender 40 anthro 40 ",
+                    "group 38 english_text 36 smile 36 simple_background 35 ",
+                    "felid 34 open_mouth 34 solo 34 clothed 32",
+                )
+                .into(),
+                related_tags_updated_at: "2021-12-11T15:59:15.118-05:00".parse().ok(),
+                category: Category::General,
+                is_locked: false,
+                created_at: "2020-03-05T05:49:37.994-05:00".parse().unwrap(),
+                updated_at: "2021-12-11T15:59:15.118-05:00".parse().unwrap(),
+            }),
+            Ok(Tag {
+                id: 12053,
+                name: "sefeiren".into(),
+                post_count: 1581,
+                related_tags: concat!(
+                    "conditional_dnp 300 sefeiren 300 female 222 male 211 claws 176 ",
+                    "anthro 173 genitals 165 mammal 165 paws 161 hindpaw 151 hi_res 150 ",
+                    "feral 149 fur 142 tongue 142 nude 131 duo 122 open_mouth 122 ",
+                    "scalie 117 bodily_fluids 116 frisky_ferals 113 horn 112 solo 112 ",
+                    "penis 108 text 104 blush 101",
+                )
+                .into(),
+                related_tags_updated_at: "2022-04-06T06:21:26.712-04:00".parse().ok(),
+                category: Category::Artist,
+                is_locked: false,
+                created_at: "2020-03-05T05:49:37.994-05:00".parse().unwrap(),
+                updated_at: "2022-04-06T06:21:26.713-04:00".parse().unwrap(),
+            }),
+            Ok(Tag {
+                id: 12052,
+                name: "skee".into(),
+                post_count: 2,
+                related_tags: concat!(
+                    "male 2 solo 2 skee 2 simple_background 2 shorts 1 bulge 1 ",
+                    "clothing 1 genital_outline 1 genitals 1 hi_res 1 huge_balls 1 ",
+                    "huge_penis 1 hyper 1 hyper_balls 1 hyper_genitalia 1 ",
+                    "hyper_penis 1 limitedvision 1 lizard 1 anthro 1 multi_genitalia 1 ",
+                    "multi_penis 1 penis 1 penis_outline 1 reptile 1 scalie 1",
```
can we have less... explicit mock data 😅

playfulkittykat replied:
Oh god I'm so sorry! I didn't even read the actual content. I'll get this cleaned up.

nasso commented on src/client.rs:
```diff
@@ -56,6 +60,43 @@ pub(crate) type QueryFuture = Box<dyn Future<Output = Result<serde_json::Value>>
+impl fmt::Display for Cursor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Page(p) => write!(f, "{}", p),
+            Self::Before(p) => write!(f, "b{}", p),
+            Self::After(p) => write!(f, "a{}", p),
```
```suggestion
            Self::Page(p) => write!(f, "{p}"),
            Self::Before(p) => write!(f, "b{p}"),
            Self::After(p) => write!(f, "a{p}"),
```

playfulkittykat replied:
Should I also add `rust-version = "1.58"` to `Cargo.toml`?

nasso replied:
oh maybe, i didn't check what the msrv is

playfulkittykat replied:
Looks like it'd be around `1.70` because of some dependencies.

nasso replied:
yeah that seems okay to be honest?

nasso commented on src/post.rs:
```diff
@@ -386,9 +377,9 @@ impl<'a> Stream for PostSearchStream<'a> {
-                            SearchPage::Page(i) => format!("{}", i),
-                            SearchPage::BeforePost(i) => format!("b{}", i),
-                            SearchPage::AfterPost(i) => format!("a{}", i),
+                            Cursor::Page(i) => format!("{}", i),
+                            Cursor::Before(i) => format!("b{}", i),
+                            Cursor::After(i) => format!("a{}", i),
```
```suggestion
                            Cursor::Page(i) => format!("{i}"),
                            Cursor::Before(i) => format!("b{i}"),
                            Cursor::After(i) => format!("a{i}"),
```

nasso reviewed (commented):
pretty cool! looking mostly good to me, just minor nitpicks (also minor conflicts)

regarding the mocked data (in tests and json files), i'd prefer using SFW data, please 😅

playfulkittykat committed:
Replace target_arch with target_family

playfulkittykat committed:
At least build for wasm32 in GitHub Actions

playfulkittykat committed:
Set minimum rust version

playfulkittykat committed:
Unignore Cargo.lock

https://doc.rust-lang.org/nightly/cargo/faq.html#why-have-cargolock-in-version-control

playfulkittykat commented:
I've added `Cargo.lock` to the repository (see [here](https://doc.rust-lang.org/nightly/cargo/faq.html#why-have-cargolock-in-version-control) for the general reasoning) specifically because I've added 1.70 to GitHub Actions, and if a dependency changes its `rust-version`, that would cause CI to randomly break.

nasso commented:
aaa i gotta review this i forgot woops

playfulkittykat commented:
I bungled this branch up in my last rebase. Give me a bit to clean it up!

playfulkittykat commented:
Actually, this one still looks fine. It's my local branch that's destroyed :rofl:
